### PR TITLE
Improve module restrict hints

### PR DESCRIPTION
**Pull request by lierdakil**
I'm not updating the readme for now, since I want some general feedback before doing any more work.

- Add new fields to module restrict config:
    - asRequired: Bool; if true, `as` alias is required. Ignored if `as`
      is empty.
    - ~~qualified: either 'always', 'never', 'exhaustive' or 'ignore';
      should the module be qualified? 'exhaustive' only requires
      qualification if there's no explicit import list, e.g.
      `import Foo (a, b, c)` is fine, but `import Bar hiding (d)` and
      `import Baz` will generate a hint.~~
    - importStyle: one of qualified, unqualified, explicit, explicitOrQualified, ignore.
      The preferred import style. `explicitOrQualified` accepts both `import Foo (a,b,c)`
      and `import qualified Foo`, but not `import Foo` or `import Foo hiding (x)`.
      `explicit` is basically the same, but doesn't accept `import qualified`
      (I'm not sure if it's very useful tbh). `qualified` and `unqualified` do not care
      about the import list at all.
    - qualifiedStyle: either 'pre', 'post' or 'ignore'; how should the
      module be qualified?

- Wildcard match handling is extended, such that restrictions are
  unified between wildcard and specific matches. With the new
  fields, the more specific option takes precedence. The list fields are
  merged. With multiple wildcard matches, the precedence between them is
  not guaranteed (but in practice, names are sorted in the
  reverse lexicograpicall order, and the first one wins -- which
  hopefully means the more specific one more often than not)

  This hopefully explains the reason for 'ignore' option: we can
  override the wildcard match with it.

  The primary motivation for this is to be able to specify the default
  pre/post style with something like

  ```
  - {name: ['**'], qualifiedStyle: post}
  ```

  But it does also add a fair bit of flexibility for writing rules (at the cost of some complexity).

  If there's a better way to do this, let me know.

If the same module is specified multiple times, for the new fields, only
the first definition will take effect.

If you're wondering why `asRequired` is a `Bool`, and not `always/never/ignore`, that's because `false` is `ignore` here, and there's no `never` option. One can already effectively forbid aliasing by specifying something like `name: Data.Map, as: Data.Map`.

### CODE REVIEW

**Comment by lierdakil** [comment]
Side note: could also tackle https://github.com/ndmitchell/hlint/issues/769 here to some extent, e.g. by adding `qualified: 'exhaustive'` option, which would generate the hint only if the import doesn't have an explicit import list.

File: .hlint.yaml
Comment (zliu41) [inline .hlint.yaml:31]: This file should not be used for testing. You can create a separate yaml file under `data`, and use it in a `test` file under `tests`. See `data/wildcard.yaml` and `tests/wildcard.test` for an example.

File: .hlint.yaml
Comment (zliu41) [inline .hlint.yaml:32]: I'm not sure the meaning of `qualified: exhaustive` is immediately obvious. Perhaps `import_style: qualified / explicit / qualified_or_explicit`?

**Review by zliu41 — COMMENTED** [review]
I like this - rules for enforcing import styles can be useful for many Haskell projects.

File: .hlint.yaml
Comment (lierdakil) [reply .hlint.yaml:32]: :thinking: with the proposed `import_style` we lose the ability to just qualification without requiring an explicit import list. That said, I don't know if it's particularly useful.

Comment (lierdakil) [reply .hlint.yaml]: I've made the changes. Not sure if `unqualified` and `explicit` are particularly useful, but maybe there's a use case for those I didn't think of.

**Review by zliu41 — COMMENTED** [review]
This looks good to me. I've always wanted to use HLint to enforce qualified or explicit import, and this does the job. @ndmitchell could you review this?

File: .hlint.yaml
Comment (zliu41) [reply .hlint.yaml]: > unqualified

I'd suggest removing this - "This module must be imported unqualified!" says no one, ever :smile:

Comment (lierdakil) [reply .hlint.yaml]: > says no one, ever

I would avoid cutting obvious functionality just because you can't think of a use case from the top of your head, it ultimately only leads to user frustration. It's not particularly costly to implement, and I'd like to have the option on the off-chance this comes in handy. One example where I imagine this might be somewhat useful is, say, custom prelude imports, which you might want to always be unqualified, but not necessarily always explicit.

**Comment by ndmitchell** [comment]
This conflicts with #1342, which renamed it to wildcard-imports.test and rewrote most of that file. Can you reapply there?

File: src/Config/Type.hs
Comment (ndmitchell) [inline src/Config/Type.hs:129]: I guess Ignore means "I don't care" - would `ImportStyleUnrestricted` be a better name?

File: src/Config/Type.hs
Comment (ndmitchell) [inline src/Config/Type.hs:135]: Ditto, QualifiedStyleUnrestricted?

Although I wonder how useful qualified style is? There is a GHC warning to require post qualified, so wouldn't people just use that? Although I'm not against it.

File: src/Config/Type.hs
Comment (ndmitchell) [inline src/Config/Type.hs:145]: Why Alt rather than just `Maybe QualifiedStyle`?

File: src/Config/Type.hs
Comment (ndmitchell) [reply src/Config/Type.hs:129]: That said, if we want ignore as the API for users in YAML, this seems fine. I think I might be ambivalent

File: src/Config/Yaml.hs
Comment (ndmitchell) [inline src/Config/Yaml.hs:332]: Would be nice to drag the matching of a list of explicit strings into a enum into a helper function, taking a list of string and value pairs. That way we could guarantee the parseFail suggested list was always perfectly correct

**Review by ndmitchell — APPROVED** [review]
Sorry it took so long to review! Looks good to me. For Alt, happy either way. For Ignore vs unrestricted, I'll let you decide, whatever you prefer. The tests need fixing after the previous diff merge, but once they are good I'll merge.

File: .hlint.yaml
Comment (ndmitchell) [reply .hlint.yaml]: I actually quite like saying modules must be imported unqualified. Every time I see `import Data.List(sort)` I reply with just don't bother listing explicit imports. So I think it's possibly something I'd use 😄 .

File: src/Config/Type.hs
Comment (lierdakil) [reply src/Config/Type.hs:135]: As I outlined in the OP, the primary reason for having this is not as much to have a warning (as you say, GHC already has a warning for this), as to be able to specify the default for HLint's hints. It would be pretty annoying if HLint says to use `import qualified X` and then GHC complains that it should be `import X qualified` instead, I think. And since we already have the option, adding a check is literally two lines of code.

File: src/Config/Type.hs
Comment (lierdakil) [reply src/Config/Type.hs:145]: Because I want to use `sconcat` for `RestrictedItem`s. Granted, I could define `Semigroup RestrictedItem` in terms of `<|>`, but defining `<>` in terms of `<|>` feels a bit weird to me.

**Comment by lierdakil** [comment]
Okay, I've rebased and applied some suggested changes (see https://github.com/ndmitchell/hlint/commit/e215ae33ea008854666c2dc654f197125e189b59), and also updated the README.

**Comment by ndmitchell** [comment]
Thanks for all your changes, looks great

**Pull request merged**
